proposed/MPID Messaging.md

```diff
@@ -0,0 +1,451 @@
+pub const MAX_BODY_SIZE: usize =
+    ::routing::structured_data::MAX_STRUCTURED_DATA_SIZE_IN_BYTES - 512 - MAX_HEADER_METADATA_SIZE;
+pub const MAX_INBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+pub const MAX_OUTBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+```
+
+### `MpidHeader`
+
+```rust
+pub struct MpidHeader {
+    sender_name: ::NameType,
+    guid: [u8; 16],
+    metadata: Vec<u8>,
+    signature: ::sodiumoxide::crypto::sign::Signature,
+}
+```
+
+The `sender` field is hopefully self-explanatory.  The `guid` allows the message to be message to be uniquely identified, both by receivers and by the manager Vaults which need to hold the messages in a map-like structure.
+
+The `metadata` field allows passing arbitrary user/app data.  It must not exceed `MAX_HEADER_METADATA_SIZE` bytes.
+
+### `MpidMessage`
+
+```rust
+pub struct MpidMessage {
+    header: ::MpidHeader,
+    recipient: ::NameType,
+    body: Vec<u8>,
+    recipient_and_body_signature: ::sodiumoxide::crypto::sign::Signature,
+}
+```
+Each `MpidMessage` instance only targets one recipient.  For multiple recipients, multiple `MpidMessage`s need to be created in the [Outbox][3] (see below).  This is to ensure spammers will run out of limited resources quickly.
+
+## Outbox
+
+This is a simple data structure for now and will be a hash map of serialised and encrypted `MpidMessage`s.  There will be one such map per MPID (owner), held on the MpidManagers, and synchronised by them at churn events.
+
+This can be implemented as a `Vec<MpidMessage>`.
+
+## Inbox
+
+Again this will be one per MPID (owner), held on the MpidManagers, and synchronised by them at churn events.
+
+This can be implemented as a `Vec<(sender_name: ::routing::NameType, sender_public_key: ::sodiumoxide::crypto::sign::PublicKey, signed_header: Vec<u8>)>` or having the headers from the same sender grouped: `Vec<(sender_name: ::routing::NameType, sender_public_key: ::sodiumoxide::crypto::sign::PublicKey, headers: Vec<signed_header: Vec<u8>>)>` (however this may incur a performance slow down when looking up for a particular mpid_header).
+
+## Messaging Format Among Nodes
+
+Messages between Clients and MpidManagers will utilise [`::routing::structured_data::StructuredData`][5], for example:
+```rust
+let sd = StructuredData {
+    type_tag: MPID_MESSAGE,
+    identifier: mpid_message_name(mpid_message),
+    data: ::utils::encode(mpid_message),
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![sender_public_key],
+    previous_owner_signatures: vec![]
+}
+```
+
+When handling churn, MpidManagers will synchronise their outboxes by sending each included MpidMessage as a single network message.  However, since MpidHeaders are significantly smaller, inboxes will be sent as a message containing a vector of all contained headers.
+
+## Message Flow
+
+![Flowchart showing MPID Message flow through the SAFE Network](MPID%20Message%20Flow.png)
```

Comment:
I was seeing this as the Client sends `Post(Online)` message and then gets each message as and when it can be retrieved by its managers from the network.  If the client is already connected when a new message is created for it, it will just receive the message.  So I don't think we need an explicit GetMessage or GetAllMessages for Client to own inbox.